### ISSUE TITLE
The stability of the disk pgen with a interacting planet

### ISSUE THREAD

**Issue by LeonKLK** [PrincetonUniversity/athena-public-version#65]
Dear all,

We are final year students studying the interaction on the accretion disk by planets. What we are trying to do, is to add a revolving planet(1 Jupiter mass) into a disk, allowing the disk and planet interact with each other. We want to take out simulation up to 5000 orbits. However, we are facing two problems. The code is shown below.

```
void Planet_and_Cooling(MeshBlock *pmb, const Real time, const Real dt,
  const AthenaArray<Real> &prim, const AthenaArray<Real> &bcc, AthenaArray<Real> &cons)
{
  // MeshBlock *pmb = pmy_hydro_->pmy_block;
  // pmb is already assigned before calling this source term function
  Real src1_P, src2_P;
  Real accx, accy, accx_P, accy_P;
  Real dt2 = dt/1.91;
  Real lfdt = 3/2*dt;

  Real r_P_cube = std::pow(1.0,3); //initial r_P in cube
  Real time_P = 2*M_PI*std::pow(r_P_cube/gm0 , 0.5); //period of planet
  AthenaArray<Real> vol(pmb->ncells1);

  if (time < time_P) {
      mass_P = final_mass_P*std::pow(std::sin((PI/2)*(time/time_P)),2);
  }
  else {
      mass_P = final_mass_P;
  }

  acc1_P = -gm0/pow(r_P,2);
  acc2_P = 0;
  hdt = dt;

  for (int k=pmb->ks; k<=pmb->ke; ++k) {
// #pragma omp parallel for schedule(static)
    for (int j=pmb->js; j<=pmb->je; ++j) {
// #pragma simd
    pmb->pcoord->CellVolume(k, j, pmb->is, pmb->ie, vol);
      for (int i=pmb->is; i<=pmb->ie; ++i) {
        Real den = prim(IDN,k,j,i);
        Real rad, phi, z;
        GetCylCoord(pmb->pcoord,rad,phi,z,i,j,k);
        Real theta = phi - phi_P; // rel. location of the planet
        //calculate the force on disk element
        soft = 0.6*r_P*std::pow(mass_P/(3*mass_S),1.0/3); // soft = k*R_H
        Real dist = std::sqrt(rad*rad + r_P*r_P - 2.0*rad*r_P*std::cos(theta));
        Real dist_soft = std::sqrt(rad*rad + r_P*r_P - 2.0*rad*r_P*std::cos(theta) + std::pow(soft,2)); // distance to the planet with soften parameter in the G-potential
        Real alpha = std::acos((r_P*r_P+dist*dist-rad*rad)/(2.0*dist*r_P)); // angle opposite to rad(radius of disk element)
        if (theta > PI){
            alpha = -alpha;
        }
        // acceleration in r and phi
        // projection on the vector r
        accx = mass_P/std::pow(dist_soft,3.0)*(dist*cos(2*M_PI-alpha)); // for r so far away, force at x-dir is +ve, for r very small, force at x-dir is-ve
        accy = mass_P/std::pow(dist_soft,3.0)*(dist*cos(1.5*M_PI-alpha)); // the angles inside come dot profuct between vector r and vector x(and vector y)
        acc1 = cos(theta)*accx + sin(theta)*accy; // rotate the force back to the cylindrical coordinate
        acc2 = -sin(theta)*accx + cos(theta)*accy;

        // sources to the radial and angular momentum eqs
        src1 = dt*den*acc1;
        src2 = dt*den*acc2;
        // update
        cons(IM1,k,j,i) += src1;
        cons(IM2,k,j,i) += src2;
        if (NON_BAROTROPIC_EOS) {
            cons(IEN,k,j,i) += 0.5*(SQR(src1)+SQR(src2))/den;
        }
        // calculate the force on planet
        //notice that the direction of vector dist is reversed, so we don't need a -ve sign in front of acc_P
        soft_in = 1 - std::exp(-dist*dist/std::pow(soft,2));
        accx_P = -den*vol(i)/std::pow(dist_soft,3.0)*(dist*cos(2*M_PI-alpha));
        accy_P = -den*vol(i)/std::pow(dist_soft,3.0)*(dist*cos(1.5*M_PI-alpha)); //*vol(i)
        acc1_P += soft_in*accx_P;
        acc2_P += soft_in*accy_P;
        // short description: the above calculate the acceleration by the planet on the elements of the disk
        // and add the acceration according to the direction of vector r and vector phi of each element
        // note that the calculation is using theta, which means we use the planet as a reference

      }
    }
  }

// for non-interacting planet
//  phi_P = omega_P*time;
//  if (phi_P > TWO_PI){
//      phi_P = phi_P - TWO_PI;
//  }

  //LF
  //For the first interaction, the above is computing a_i
  //in order to compute v_i+1, we require a_i+1, which is computed in the next iteration
  //that's why we save the i-iteraction's acc_P as old_acc_P for the next iteration's v_i+1

  //LF integrator
  old_r_x = r_P*cos(phi_P);
  old_r_y = r_P*sin(phi_P);

  old_v_x = vr_P*cos(phi_P) - vphi_P*sin(phi_P);
  old_v_y = vr_P*sin(phi_P) + vphi_P*cos(phi_P);

  a_x = acc1_P*cos(phi_P) - acc2_P*sin(phi_P);
  a_y = acc1_P*sin(phi_P) + acc2_P*cos(phi_P);

  // This is the calculation of v_i, since we need a_(i+1) for v_i, so we hold the calculation of v_i
  // at the initial time.
  // x_(i+1) = x_i +v_i*dt + 0.5*a_i*dt*dt
  // v_(i+1) = v_i + 0.5*(a_i+a_(i+1))*dt*dt

  // notice that for x1 we need v_0 only
  // so we don't need to compute v at this moment
  if (time > lfdt){
      v_x = old_v_x + 0.5*(old_a_x +  a_x)*(dt2);
      v_y = old_v_y + 0.5*(old_a_y +  a_y)*(dt2);
  } else {
      v_x = old_v_x;
      v_y = old_v_y;
  }

  r_x = old_r_x + v_x*(dt2) + 0.5*a_x*(dt2)*(dt2);
  r_y = old_r_y + v_y*(dt2) + 0.5*a_y*(dt2)*(dt2);

  old_a_x = a_x;
  old_a_y = a_y;

  r_P = sqrt(r_x*r_x + r_y*r_y);
  phi_P = std::atan2(r_y, r_x);

  vr_P = v_x*cos(phi_P) + v_y*sin(phi_P);
  vphi_P = -v_x*sin(phi_P) + v_y*cos(phi_P);

  // Rmk: the above leapfrog integrator is using a fixed coordinate setted by athena++,
  // that's why we rotate the system using phi_P

  //since we CANNOT enroll two source term function, so we combine the source functions together
  //cooling function
  Real omega;
  Real tau_r;

  for (int k=pmb->ks; k<=pmb->ke; ++k) {
      for (int j=pmb->js; j<=pmb->je; ++j) {
          pmb->pcoord->CellVolume(k, j, pmb->is, pmb->ie, vol);
          for (int i=pmb->is; i<=pmb->ie; ++i) {
              Real rad, phi, z;
              GetCylCoord(pmb->pcoord,rad,phi,z,i,j,k);
              Real den = prim(IDN,k,j,i);
              Real poverr;
              omega = prim(IVY,k,j,i)/rad;
              tau_r = 2*M_PI*tau/omega;
              poverr = p0_over_r0*std::pow(rad/r0, -1)/gamma_gas;
              Real new_pres = prim(IPR,k,j,i) - dt/tau_r*(prim(IPR,k,j,i) - poverr*den);
              cons(IEN,k,j,i) += (new_pres-prim(IPR,k,j,i))/(gamma_gas-1);
          }
      }
  }

  return;
}
```

1: Inside our user-defined source function, there are mainly 3 parts: effect on disk by the planet, effect on the planet by disk(Leapforg integrator) and the cooling function. The first problem comes from the the position of planet. Using the unit we have(G*M_Sun=1, orbital radius=1 AU) , we expect the planet to finish one orbit at 6.28 unit of time(notice that the total mass of disk is only 0.002 solar mass so it won't affect much), meaning for 100 orbit, the `tlim` should be set as 628. However, we found that the value of period is not same as expected, especially it changes with the choice of time-integrator(rk2, rk3).

![dt](https://user-images.githubusercontent.com/66102343/97099235-d3a16100-16c1-11eb-8331-65a084e2312f.PNG)

We print out the value of `dt` inside the our source function(denoted as `hdt`), we found that the values are not the same as `dt` we expected(`hdt` and `dt` have a fixed ratio), thus causing the problem. Right now our solution on this is to create a new variable(`dt2`) and adjust using the ratio. Is there any better way to do it? 

2: The second problem is quite straight forward, unstable. Even if we fixed the above problem, the code would crash eventually.
We have tried different combination of resolution, CFL number and time-integrator, all combinations fail to finish the simulation. 

![rec 3](https://user-images.githubusercontent.com/66102343/97099329-bde06b80-16c2-11eb-9a2d-47c6264ae949.PNG)

As you can see above, the value of `dt` were stable, then the value of `dt` increased significantly after the crashing and the simulation "finished". Below shows the typical crashing scenario, with the values of prim variable being NAN at some time step.

![nan](https://user-images.githubusercontent.com/66102343/97099376-2deef180-16c3-11eb-8a81-14ba118a2dd8.PNG)

We even tried using a non-interacting planet(disk won't affect the planet, so the planet revolves independently), and the simulation still crashed.

Did we misuse your code? Or are there any bugs we missed in the above source function?
It would be great if you can give us some insight and help. Thank you.

Leon

**Comment by c-white** [comment]
Just some immediate thoughts that occurred to me as I scanned the code:

- `Real lfdt = 3/2*dt;` is equivalent to `lfdt = dt;`.
- The code looks like it will only work in serial. If this is run in parallel I would expect it to break.
- Mixing integrators is very tricky. It might help to step back and diagram what the Athena++ integrators are doing to figure out how a leapfrog integrator could fit in. For example, VL2, RK2, and RK3 take 2, 2, and 3 forward Euler substeps per timestep, respectively. The `dt` given to the source function is not the global `dt` for the full step, but rather some fraction of this to use in the substep. These fractions are 1/2 and 1 for VL2; 1 and 1/2 for RK2; and 1, 1/4, and 2/3 for RK3. So when you print `hdt` and it shows 2/3 the value of the global `dt`, that just means you were using RK3 and only the third of three substeps is being output, since outputs are only made at the end of the full step. Note the substep lengths do not add up to 1. I'm guessing the planet seems to move 50% faster than it should for VL2/RK2, and nearly twice as fast for RK3.
- This applies to more than just timesteps. What state the `prim` and `cons` arrays are in varies from substep to substep -- they are necessarily not always the values left at the end of the previous substep. The planet's location, velocity, and acceleration at the beginning of the substep likewise should not be taken to be the same as at the end of the previous substep.
- An easier integrator to work with is RK1, which is just a single forward Euler substep.
- As for crashing, that's probably best investigated by making 2D plots of various quantities. In all likelihood there is a single location where the density or pressure became negative. Is it happening close to the planet, at a grid boundary, everywhere?

**Comment by LeonKLK** [comment]
Hi Chris, thank you for the reply.

- For the initial time step, we don't have to update the planet's velocity, that's why we set up such `if-else` statement.  Since I am not sure how many times my source function is executed, in order to make sure my planet is updated with the correct position, it's more like a precaution setting. If setting `lfdt` is meaningless, I think I can simply remove it.

- If you are talking about running it with MPI, we never tried running it in a parallel way so it's not the reason of crashing.

- This is a very important information for us! If we use `dt` in my leapfrog integrator(for rk2), the planet would finish one orbit around `t=4.188` which is about 33.3% shorter than expected. In terms of orbital velocity, just like what you expected, the value is 50% higher.

- Since our project followed a previous paper([https://arxiv.org/pdf/1508.00903.pdf](url))(using PLUTO, another simulation code), we prefer having the same setting in order to replicate their result as our first step. Of course we can try using RK1 or any other possible options, but are there any ways to use the LF integrator with the higher-order time integrator?

- From the tab file or the vtk file, I don't see any negative density or pressure. I have tried output tab files for every global `dt`, for most of the time, the pressure and density of whole disk dropped to the default `dfloor` and `pfloor`, and the velocities became NAN, and the previous tab file seems to have no anomaly. Below are some vtk files before and after crashed, the setting is 256x256 in cylindrical coordinate, CFL=0.2, and it crashed at t=68.4175. We set the r-direction boundary to be reflective and phi-direction to be periodic. 

P.S. Here is a output movie from VISIT, you may take a look (just notify me if the link is not working)
[https://drive.google.com/file/d/1NmyvoM5vX2iM3CUuf11R563dSVqnIyN0/view?usp=sharing](url)

![1](https://user-images.githubusercontent.com/66102343/97143800-539cf900-179e-11eb-8286-bc30522b199b.PNG)
![2](https://user-images.githubusercontent.com/66102343/97143806-5566bc80-179e-11eb-9a61-ed629fe08db2.PNG)
![3](https://user-images.githubusercontent.com/66102343/97143810-55ff5300-179e-11eb-8ed5-b392dcdb1b36.PNG)



In the meantime, I would try to use RK1 first to see whether it would be more stable. I noticed that there are some related researches done by others and they used Athena++ (for example: https://arxiv.org/pdf/1911.01530.pdf), so I guess there should be a way out, just we haven't found it yet. Again, thank you very much for the information provide. it's much appreciated.

**Comment by c-white** [comment]
For my first point, I only meant that `3/2*dt` will evaluate to `1*dt`, since `3/2 == 1` in integer arithmetic. You probably wanted `3.0/2.0*dt`.

As for mixing integrators, the easiest case (beyond RK1) is probably VL2. Here there are two substeps, the predict step and the correct step. In the predict step, you are given `time` t_n; `dt` equal to dt_full_step / 2; `prim` at t_n; and `cons`, which is at t_{n+1/2} = t_n + dt_full_step / 2, except for the source terms which it is your responsibility to add at this point, using a step of `dt`. In the correct step, you are given `time` t_{n+1/2}; `dt` equal to dt_full_step; `prim` at t_{n+1/2}, which is consistent with what you set `cons` to during the predict step; and `cons`, which is at t_{n+1} = t_n + dt_full_step, again except for the source terms you are presently adding, again with `dt` (which is different from before).

Suppose during the predict step you have the planet position `x_new` at t_n and velocity `v` at t_{n-1/2}, as well as `dt_full_previous`, which is the value of `dt` passed to the function during the previous call (the correct step of the previous time step). You can make the following calculations:

1. Calculate planet acceleration `a` == a_n from `x_new` == x_n and `prim` == P_n.
2. Update conserved quantities `cons` from C*\_{n+1/2} to C_{n+1/2} with `a` and `dt`.
3. Update velocity `v` from v_{n-1/2} to v_{n+1/2} with a_n and a change in time `dt + dt_full_previous/2.0`.
4. Store `x_old = x_new`.
5. Calculate `x_new` == x_{n+1/2} from x_n, v_{n+1/2}, and `dt`.

Now consider the correct step. You have `x_old` at t_n, `x_new` at t_{n+1/2}, and `v` at t_{n+1/2}. Make the following calculations:

1. Calculate `a` == a_{n+1/2} from `x_new` and `prim` == P_{n+1/2}.
2. Update `cons` from C*\_{n+1} to C_{n+1} with `a` and `dt`.
3. Calculate `x_new` == x_{n+1} from `x_old`, `v`, and `dt`.
4. Store `dt_full_previous = dt`.

This is entirely untested, but it outlines how I would go about mixing LF for the planet with VL2 for the gas. RK2 would be different but perhaps no more complicated, while RK3 could potentially be worse. A quick hack to know which substep you are in is just to use a static variable that toggles every time the function is called.

As a side note, cooling functions can also be a bit tricky. The explicit method you are using can work, but it can be dangerous, since there is no guarantee the pressure will not become negative in a single step. This might explain how everything crashes, especially since a single bad might propagate to the entire grid in a single step due to the fact that every cell influences the planet, which then influences every cell. You can check with some print statements to see if you ever have `prim(IPR,k,j,i)` positive but `new_pres` negative.

The safer strategy is a locally implicit method. In this case, ignore `prim` entirely. If you want a gas quantity that's normally found in `prim`, recompute it from `cons`. Now your goal is, given pre-cooled `cons` C_pre, replace it with post-cooled C_post that accounts for a step of `dt` (whatever `dt` is passed into the function). That is:

1. Calculate rho, p_pre, and v from C_pre.
2. Solve (p_post - p_pre) / dt = f(rho, p_post, ...) for p_post given whatever your cooling function f looks like.
3. Reset C_pre using rho, p_post, and v.

**Comment by LeonKLK** [comment]
Thank you for the reminder about integer arithmetic, now it's adjusted.

About the mixing integrator, I think it may not be a good idea to mix things like this because the logic seems way to complicated and it maybe hard for debugging, not to mention the mixing `RK3` or `RK4`. Instead of mixing, we came up with another way out
```
  dt_step = 1;
  if (dt_step % 2 == 1){
    old_r_x = r_P*cos(phi_P);
    old_r_y = r_P*sin(phi_P);

    old_v_x = vr_P*cos(phi_P) - vphi_P*sin(phi_P);
    old_v_y = vr_P*sin(phi_P) + vphi_P*cos(phi_P);

    a_x = acc1_P*cos(phi_P) - acc2_P*sin(phi_P);
    a_y = acc1_P*sin(phi_P) + acc2_P*cos(phi_P);

    // This is the calculation of v_i, since we need a_(i+1) for v_i, so we hold the calculation of v_i
    // at the initial time.
    // x_(i+1) = x_i +v_i*dt + 0.5*a_i*dt*dt
    // v_(i+1) = v_i + 0.5*(a_i+a_(i+1))*dt*dt

    // notice that for x1 we need v_0 only
    // so we don't need to compute v at this moment
    if (time > dt){
        v_x = old_v_x + 0.5*(old_a_x +  a_x)*(dt);
        v_y = old_v_y + 0.5*(old_a_y +  a_y)*(dt);
    } else {
        v_x = old_v_x;
        v_y = old_v_y;
    }

    r_x = old_r_x + v_x*(dt) + 0.5*a_x*(dt)*(dt);
    r_y = old_r_y + v_y*(dt) + 0.5*a_y*(dt)*(dt);

    old_a_x = a_x;
    old_a_y = a_y;

    r_P = sqrt(r_x*r_x + r_y*r_y);
    phi_P = std::atan2(r_y, r_x);

    vr_P = v_x*cos(phi_P) + v_y*sin(phi_P);
    vphi_P = -v_x*sin(phi_P) + v_y*cos(phi_P);
  }
  dt_step += 1;
```
we simply set up a global variable `dt_step`, since we know how many substeps for each integrator and which substep corresponds for the `full_dt`, we can use this `if-statement` to ensure the planet-position update would only be execute with `full_dt`. In this way, our LF integrator works like a independent integrator. (We could write it in a more general way for different integrators, for now we use this for `RK2` first.)

As for the cooling function, we tested two things:

1: `tau_r` controls the intensity of, we use `tau_r=0.01` for most of our simulations. However, when we changed the value to `tau_r=10`(slower cooling), the crashing time changed from `t=17.974` to `t=62.6456`. 
Right now at the moment I am running the simulation without the cooling function, the simulation is stilling running at `t=109` now. If the simulation can finish without crashing, probably the crashing problem comes from the cooling function itself.

2: About ignoring `prim` entirely, for rho and v, I can simply calculate it using cons(rho) and cons(momentum).
For prim(pressure), do you mean I should replace prim(pressure) from cons(energy) using the following relation?
```
prim(IPR,k,j,i) = cons(IEN,k,j,i)*(gamma_gas-1)
```
I tried using the above relation to replace prim(pressure) whenever prim(pressure) appears in cooling(but not rho and v), e.g.
```
for (int k=pmb->ks; k<=pmb->ke; ++k) {
      for (int j=pmb->js; j<=pmb->je; ++j) {
          pmb->pcoord->CellVolume(k, j, pmb->is, pmb->ie, vol);
          for (int i=pmb->is; i<=pmb->ie; ++i) {
              Real rad, phi, z;
              GetCylCoord(pmb->pcoord,rad,phi,z,i,j,k);
              Real den = prim(IDN,k,j,i);
              Real poverr;
              omega = prim(IVY,k,j,i)/rad;
              tau_r = 2*M_PI*tau/omega;
              poverr = p0_over_r0*std::pow(rad/r0, -1)/gamma_gas;
              //Real new_pres = prim(IPR,k,j,i) - dt/tau_r*(prim(IPR,k,j,i) - poverr*den);
              //cons(IEN,k,j,i) += (new_pres - prim(IPR,k,j,i))/(gamma_gas-1);
              cons(IEN,k,j,i) += (-dt/tau_r*(cons(IEN,k,j,i)*(gamma_gas-1)-poverr*den))/(gamma_gas-1);
          }
      }
  }
```
The problem is, the simulation crashed within its first orbit and the value of pressure and momentum are quite different.
![prim](https://user-images.githubusercontent.com/66102343/98065134-dacd2980-1e8e-11eb-910d-c2df7596dc8a.PNG)
For the upper one, the result comes from the last line of the above code. You can see it crashed at `t=2.79`. The lower one comes from the original cooling function(the two comment-out lines from above). You can see that both the momentum and kinetic energy are quite different in order. Also from the tab files of the upper one simulation, the pressures of all grids drop to `pfloor` very quickly. 
I am not sure what happened since I just gave a quick trial.  I will try to figure out the cooling function with `prim` ignored asap after my current running simulation is finished.

UPDATE: Simulation(512x512, CFL=0.4) without cooling still crashed at `t=152.7` which is the farthest we can go. Meaning no cooling/ slower cooling can delay the crashing time.

**Comment by LeonKLK** [comment]
In past few days I had been running tests for the code and I found one bug in my code.

![rk1](https://user-images.githubusercontent.com/66102343/99075643-c9092600-25f4-11eb-93ed-faa267230a09.PNG)

In the above vtk the colour is in log-sacle(to emphasize the contrast), we can see that near the position of planet, when it starts to revolving below x-axis(phi > PI), the drop to the `pfloor` for pressure shows up, and this only happens between (2*PI > phi > PI). The problem comes from the function `atan2` for calculating position of planet, as its outputs range from -PI to PI, thus it causes error for this line of code
`Real theta = phi - phi_P;`
Right now the problem is fixed, however at some time we can see a grid drop to `pfloor` again all the sudden near the inner boundary and then return to "normal" as it evolves.

I have ran my code with resolution 256x512, `RK1` and `cfl=0.4`(with cooling). The code crashed at `t=109`(kind of an improvement).

Now there are two new problems, in one test(256x512, `RK3`, `cfl=0.4`) it crashed at `t=1.54`(similar to `RK2), I tried to produce a large number of vtk to see what happened and I found the following 

![rk3](https://user-images.githubusercontent.com/66102343/99076719-8f391f00-25f6-11eb-93db-6f3baa3ea6e3.PNG)

Just before it crashed we can see there are some cross structures appeared(drop to `dfloor` and 'pfloor`), I am sure this is not related to the bug mentioned the above since the planet is still above x-axis(`r=1.00138e+00 AU` and `phi=1.53952e+00 rad`)

Right now I am running the code with 256x512, `RK2`, `cfl=0.5`, the code didn't crash that quick at `t=1.54`, I would update you if it finishes(or crashes).

Summary for my questions:
1: Is there any better way to write my cooling function?
I tried what you said, to calculate `prim(IPR,k,j,i)` from `cons(IEN,k,j,i)`, assuming my code/modification is correct, the result is quite weird and crashed even quicker.

2: The cross structures mentioned above.

3: Change in `cfl` makes the code avoid crashing at the beginning.

4: The best result so far is `t=109` with cooling, it would be great if we can at least finish 100 orbits.(about `t=630`)

Sorry for so many questions from our project, your help is really appreciated.

UPDATE: Simulation(256x512, `RK2`, `cfl=0.5`) crashed at `t=173.6`, which is by far the longest time we ever achieved. Perhaps if we increase the resolution(512x512/512x1024), the simulation could go for a longer time.
